src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1495,6 +1497,161 @@ static bool extendedTypeIsOrDerivedFrom(NamedTypeSymbol hiddenContainer, NamedTy
+        /// </summary>
+        private void MergeExtensionLookupResultsHidingLessSpecificNonMethods(LookupResult current, LookupResult additional, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (!current.IsMultiViable || !additional.IsMultiViable)
+            {
+                current.MergePrioritized(additional);
+                return;
+            }
+
+            var currentSymbols = current.Symbols;
+            var additionalSymbols = additional.Symbols;
+
+            var currentCount = currentSymbols.Count;
+            var additionalCount = additionalSymbols.Count;
+            Debug.Assert(currentCount > 0);
+            Debug.Assert(additionalCount > 0);
+
+            for (int i = 0; i < additionalCount; i++)
+            {
+                var additionalSymbol = additionalSymbols[i];
+
+                removeLessSpecificNonMethodCurrentSymbols(additionalSymbol, currentSymbols, out bool isAdditionalSymbolHidden,
+                    currentCount, basesBeingResolved, ref useSiteInfo);
+
+                if (!isAdditionalSymbolHidden)
+                {
+                    currentSymbols.Add(additionalSymbol);
+                }
+            }
+
+            return;
+
+            // Determine whether we want to add the additional symbol
+            // and potentially remove some "hidden" non-method current symbols in the process
+            void removeLessSpecificNonMethodCurrentSymbols(Symbol additionalSymbol, ArrayBuilder<Symbol> currentSymbols, out bool isAdditionalSymbolHidden,
+                int currentCount, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                isAdditionalSymbolHidden = false;
+                for (int j = 0; j < currentCount; j++)
+                {
+                    var currentSymbol = currentSymbols[j];
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    if (currentSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } currentContainerUnderlying
+                        || additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying)
+                    {
+                        // We wouldn't have gathered these members if we didn't have an extended type for them.
+                        throw ExceptionUtilities.Unreachable();
+                    }
+
+                    if (additionalContainerUnderlying.Equals(currentContainerUnderlying, TypeCompareKind.ConsiderEverything))
+                    {
+                        // neither underlying type is more specific
+                    }
+                    else if (IsDerivedType(baseType: additionalContainerUnderlying, derivedType: currentContainerUnderlying,
```

Comment:
>IsDerivedType [](http://example.com/codeflow?start=29&length=13)

Do we need to handle type parameters as underlying types?

Reply:
We don't support extension member lookup on type parameters. I'll add a PROTOTYPE comment here in case we do allow it

Reply:
> We don't support extension member lookup on type parameters. I'll add a PROTOTYPE comment here in case we do allow it

Consider also adding an assert that the type symbols we are dealing with are not type parameters